src/components/FileList.tsx

```diff
@@ -237,8 +336,116 @@ export class FileList extends React.Component<IFileListProps, IFileListState> {
     await this.addFile(...this.markedFiles.map(file => file.to));
   };
 
-  updateSelectedFile = (file: Git.IStatusFile | null): void => {
-    this.setState({ selectedFile: file });
+  selectOnlyOneFile = (file: Git.IStatusFile): void => {
+    this.setState({
+      selectedFiles: [file],
+      lastClickedFile: file,
+      selectedFileStatus: file.status
+    });
+  };
+
+  toggleFile = (file: Git.IStatusFile): void => {
+    if (file.status !== this.state.selectedFileStatus) {
+      this.selectOnlyOneFile(file);
+      return;
+    }
+
+    const fileStatus = this.state.selectedFiles.find(fileStatus =>
+      filesAreEqual(fileStatus, file)
+    );
+    if (!fileStatus) {
+      this.setState({
+        selectedFiles: [...this.state.selectedFiles, file],
+        lastClickedFile: file
+      });
+    } else {
+      this.setState({
+        selectedFiles: this.state.selectedFiles.filter(
+          fileStatus => !filesAreEqual(fileStatus, file)
+        ),
+        lastClickedFile: file
+      });
+    }
+  };
+
+  removeDuplicateSelectedFiles = (): void => {
+    const selectedFiles = this.state.selectedFiles.filter(
+      (file, index, arr) => index === arr.findIndex(f => filesAreEqual(f, file))
+    );
+    this.setState({ selectedFiles });
+  };
+
+  selectFiles = (files: Git.IStatusFile[]): void => {
+    this.setState(
+      {
+        selectedFiles: [...this.state.selectedFiles, ...files]
+      },
+      () => this.removeDuplicateSelectedFiles()
+    );
+  };
+
+  handleShiftClick = (file: Git.IStatusFile): void => {
+    if (
+      !this.state.lastClickedFile ||
+      file.status !== this.state.selectedFileStatus
+    ) {
+      this.selectOnlyOneFile(file);
+      return;
+    }
+    const statusesInOrder: Git.Status[] = [
+      'unmerged',
+      'remote-changed',
+      'staged',
+      'unstaged',
+      'untracked'
+    ];
+    const partiallyStagedFiles = this.props.files.filter(
+      file => file.status === 'partially-staged'
+    );
+
+    // files that are not partially staged
+    const filesSortedByStatus: Git.IStatusFile[] = this.props.files.filter(
+      file => file.status !== 'partially-staged'
+    );
+    // add partially staged files to the array as two item (staged and unstaged)
+    partiallyStagedFiles.forEach(partiallyStagedFile => {
+      filesSortedByStatus.push({
+        ...partiallyStagedFile,
+        status: 'staged'
+      });
+      filesSortedByStatus.push({
+        ...partiallyStagedFile,
+        status: 'unstaged'
+      });
+    });
+
+    // sort the array by file status then by filename
+    filesSortedByStatus.sort((a, b) => {
+      if (a.status !== b.status) {
+        return (
+          statusesInOrder.indexOf(a.status) - statusesInOrder.indexOf(b.status)
+        );
+      } else {
+        return a.to.localeCompare(b.to);
+      }
+    });
```

Comment:
Let's rework this to avoid sorting again the `this.props.files` list.

As we forbid selection within multiple group, I suggest to filter `this.props.files` to keep only files for the current selection group (with a special conditions for _partially-staged_). Then you can look at the index in that sub array.